### PR TITLE
improve(LM rewards): uUSDwETH-DEC and YD-ETH-MAR21 week 21 liquidity mining rewards

### PR DESCRIPTION
**Pull request by chrismaree**
## Introduction
This PR adds week 22’s liquidity mining payouts for `uUSDwETH-DEC` and `YD-ETH-MAR21`. This week's payouts considered a roll between two contracts, making it somewhat more complex than a usual weekly payout. Specifically, this week contains the roll from the `uUSD-DEC` to `YD-ETH-MAR21`. During this roll there are two distinct periods: 
1) a period where only the `uUSD-DEC` receives rewards are paid: from 2020-12-21 23:00:00 UTC (11499618) to  2020-12-28 23:00:00 UTC (11525623).
2) A period where both the `uUSD-DEC` and `YD-ETH-MAR21` receive rewards, with liquidity rewards split between them: from 2020-12-25 23:00:00 UTC (11525623) to 2020-12-28 23:00:00 UTC (11545304).

For more detail on the roll, the dates and implications please see the medium post [here](https://medium.com/uma-project/uma-yield-dollar-dapp-mining-christmas-rollover-e2c07396e7f).

**Note** that the payouts for this week's rewards are **still** based on the [UMA Dev mining program](https://medium.com/uma-project/uma-announces-developer-mining-6f6fe15d5604). Based off the dev mining program, half of all rewards allocated to the uUSDwETH and YD-ETH-MAR21 contrac are allocated as liquidity mining rewards. 

This weeks developer mining rewards can be found in this PR https://github.com/UMAprotocol/protocol/pull/2359/files.

For this week's the developer mining rewards were calculated by considering the _sum_ of the uUSDwETH-DEC and YD-ETH-MAR21 pool over the duration. Then, this output was input into the `liquidity mining` after splitting the rewards pro-rata between the two distinct time periods outlined above (period with only one pool & period with the two pools.

## Payout calculation
Based on the weekly developer mining rewards, `uUSDwETH-DEC` was allocated `21415.139187657871661107`
 and `YD-ETH-MAR21` was allocated `5813.327238112296240263`. This totals `27228.4664257702`. As UMA is giving half of these to liquidity mining, this equates `27228.4664257702/2 = 13614`.

From the periods defined above, we can split these rewards into two fractions during the roll. Numerically, this equates to:
```
Period 1: (11525623-11499618)/(11545304-11499618) = 0.5692115747
Period 2: 1- 0.5692115747 = 0.4307884253
```

Note this follows the convention described in more detail in PR https://github.com/UMAprotocol/protocol/pull/1920 where the rolling logic was implemented and in PR https://github.com/UMAprotocol/protocol/pull/1929 where a previous roll was conducted.
 
Using these fractions, the total rewards allocated to the two contracts can be split up as follows:

```
Period1: 13614 * 0.5692115747 = 7749
Period2: 13614 * 0.4307884253 = 5865
```

Finally, the scripts can be run using these inputs (block numbers from periods & rewards per period). 

## Recreating the results using the scripts.

1) First, run the normal `CalculateBalancerLPRewards.js` script to calculate payouts for the `uUSD-DEC` contract. You can do this by running:
```
node ./CalculateBalancerLPRewards.js --fromBlock 11499618 --toBlock 11525623 --poolAddress="0xCce41676a4624F4A1e33a787A59d6bF96e5067bc" --tokenName="yusdeth" --week=22 --umaPerWeek=7749 --network mainnet_mnemonic
```
This will generate the weekly payout file, as you would normally. It should be named `Week_22_Mining_Rewards.json`. 

Note that the `fromBlock` here is the same block as the starting point in the weekly developer mining rewards in PR https://github.com/UMAprotocol/protocol/pull/2359/files. The `toBlock` is the block number that the roll overlap starts, December 28th, 23:00 UTC. 

2) Next, run the roll script. This script considers the two pool addresses of the two contracts `uUSD-DEC` and `YD-ETH-MAR21`. Do this by running:
```
node ./liquidity-mining/CalculateRollBalancerLPRewards.js --fromBlock 11525623 --toBlock 11545304 --pool1Address="0xCce41676a4624F4A1e33a787A59d6bF96e5067bc" --pool2Address="0x5e065D534d1DAaf9E6222AfA1D09e7Dac6cbD0f7" --tokenName="yusdeth" --rollNum=3 --umaPerWeek=5865 --synth1Address="0xD16c79c8A39D44B2F3eB45D2019cd6A42B03E2A9" --synth2Address="0x90f802C7E8fb5D40B0De583e34C065A3bd2020D8" --network mainnet_mnemonic
```
This will produce another file called `Roll_3_Mining_Rewards.json` which contains the data considering the overlap between the pools (the "roll").

Note that the `fromBlock` here is the block that the roll overlap begings. The `endBlock` is the same block as the ending point in the weekly developer mining rewards in PR https://github.com/UMAprotocol/protocol/pull/2359/files.

3) Finally, we need to join the payout files. This is done so we have one single `shareHolderPayout` which considered period 1 and period two of the two contracts & pools. To do this run:

```
node ./JoinRolledPayouts.js --week 22 --rollNum 3 --tokenName yusdeth
```

This will edit the `Week_22_Mining_Rewards.json` by combining the payouts of Period1 and Period2. it will keep all input data for prosperity.

## Script edit
A small change to `packages/affiliates/liquidity-mining/JoinRolledPayouts.js` script was required to accommodate the ordering of the roll. The previous two-phase roll (see PR https://github.com/UMAprotocol/protocol/pull/1929) had the overlap period spanning two pools at the start of the week. In this contract, it sits at the end of the week. The modification enables the joiner script to be compatible with either configuration.

### CODE REVIEW

File: packages/affiliates/liquidity-mining/yusdeth-weekly-payouts/Week_22_Mining_Rewards.json [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+  "fromBlock": 11499618,
+  "toBlock": 11545304,
+  "poolAddress": ["0xCce41676a4624F4A1e33a787A59d6bF96e5067bc", "0x5e065D534d1DAaf9E6222AfA1D09e7Dac6cbD0f7"],
+  "blocksPerSnapshot": 256,
+  "rollNum": 3,
+  "nonOverlapUma": 7749,
+  "OverlapUma": 5865,
+  "startRollBlock": 11545304,
+  "umaPerWeek": 13614,
+  "shareHolderPayout": {
```
Comment (chrismaree) [inline packages/affiliates/liquidity-mining/yusdeth-weekly-payouts/Week_22_Mining_Rewards.json:12]: This is the final shareholder payout. UMA liquidity miners are the numbers that matter & correspond to the rewards you will recive. If your address appears multiple times it's due to the roll payouts to enable auditability.

**Review by daywiss — APPROVED** [review]
Thanks for working on this Chris and leaving such detailed notes. Based on your description it makes sense to me. The numbers are pretty hard to verify though. How do you normally sanity check these things? LGTM

File: packages/affiliates/liquidity-mining/yusdeth-weekly-payouts/expiring-contract-rolls/Roll_3_Mining_Rewards.json [hunk 2]
```diff
@@ -0,0 +1,125 @@
+{
+  "rollNum": 3,
+  "fromBlock": 11525623,
+  "toBlock": 11545304,
+  "pool1Address": "0xCce41676a4624F4A1e33a787A59d6bF96e5067bc",
+  "pool2Address": "0x5e065D534d1DAaf9E6222AfA1D09e7Dac6cbD0f7",
+  "umaPerWeek": 5865,
+  "blocksPerSnapshot": 256,
```
[the remark this answers is not in the document]
Comment (chrismaree) [reply packages/affiliates/liquidity-mining/yusdeth-weekly-payouts/expiring-contract-rolls/Roll_3_Mining_Rewards.json:8]: correct.

File: packages/affiliates/liquidity-mining/yusdeth-weekly-payouts/Week_22_Mining_Rewards.json [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+    "0xd9a4693bc2d2b0246f931cd307d2bc53328cd73d": "4.046982231070218827",
+    "0x07619fbec9132d12f81b8de9e6a82e6de2589765": "0.676791064558584934",
+    "0x1d933fd71ff07e69f066d50b39a7c34eb3b69f05": "3.171819551676612828",
+    "0x46a0b4fa58141aba23185e79f7047a7dfd0ff100": "67.899490033747779219",
+    "0x5e527e59b0034efbae0e3cef30ca331a13862a84": "0.002673920757917523",
+    "0x6b244829ca823ad055c69e569f7ad2a3b64b2dc5": "2.365319986743816516",
+    "0xa9fc04a6baac104b5eaf9e91fca2df1f8d09c38c": "6.139845800837123613",
+    "0xd1b572f9528b70df1ea79456edc8250125f2d6bb": "7.534783993079717126"
+  },
+  "weeklyShareHolderPayoutBeforeRollJoin": {
```
[the remark this answers is not in the document]
Comment (chrismaree) [reply packages/affiliates/liquidity-mining/yusdeth-weekly-payouts/Week_22_Mining_Rewards.json:147]: yup! this is the payouts for period 1 where it is only the `uUSDwETH-DEC` contract.

**Comment by chrismaree** [comment]
> Thanks for working on this Chris and leaving such detailed notes. Based on your description it makes sense to me. The numbers are pretty hard to verify though. How do you normally sanity check these things? LGTM

I normally do sanity checks by looking for sponsors that I can find in both contracts and checking that they are allocated rewards correctly between the two contracts. I will also try to find a sponsor that only is in the new contract, and thus should only get rewards for the roll period. Lastly, I always check that the largest and smallest sponsors match up to what is expected from the sponsor's table. It's pretty manual but is the only real way for now.

File: packages/affiliates/liquidity-mining/yusdeth-weekly-payouts/Week_22_Mining_Rewards.json [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+    "0xed02f16a57c32a08b55d923bf4690d200722f462": "0.234505870870178531",
+    "0xef8305e140ac520225daf050e2f71d5fbcc543e7": "0.000571146191507135",
+    "0xf1ee1f3c5b4fc684d50b61c458341a73abd63b1d": "0.201764105820403436",
+    "0xf307077dd0e27a382e93f2e3d13a9c6584582332": "0.008263444275793229",
+    "0xf6f26015f0760f7a4b7f93e765bc48883daf79a5": "14.564231127968763209",
+    "0xf8a8d25049ebfaf36cf1dd7ff51ebd0777fc9b32": "40.859091614066124522",
+    "0xf94b3aa4415a14b289e329f9a75323656a2dcbc2": "2.324524535372499293",
+    "0xfc5be65e67bf016e414db1655133529033f0fa15": "0.455056809400005615",
+    "0xfdcfc1188005779ad1c70a0560fd5bbadaffebf5": "0.49548208469855303",
+    "0xffd44213718b48a371424303fe4aa2395ed4ce62": "1320.276063440137078706"
```
Comment (daywiss) [inline packages/affiliates/liquidity-mining/yusdeth-weekly-payouts/Week_22_Mining_Rewards.json:264]: this address exists in this pool as the second largest position `0xCce41676a4624F4A1e33a787A59d6bF96e5067bc` but does not exist in the second pool `0x5e065D534d1DAaf9E6222AfA1D09e7Dac6cbD0f7`, yet if i understand this right, the address earns UMA in both. Is this right?

Comment (daywiss) [reply packages/affiliates/liquidity-mining/yusdeth-weekly-payouts/Week_22_Mining_Rewards.json:264]: actually nevermind, i think i needed to refresh the page balance pool page between switching contracts 😩

**Comment by hlambur** [comment]
`uUSD-wETH` sent with txs:
https://etherscan.io/tx/0x98a4391f78caa5eb715c0d648e949eee6251c9db1622219fdd43c276b4dc08ee
https://etherscan.io/tx/0x14d1d97e6f329c171d0f54d06129e9348b2db4f991be60fb99c260bbc2e558ef

**Pull request merged**